### PR TITLE
add bazel third_party support

### PR DESCRIPTION
**Pull request by fansehep**
### What problem does this PR solve?

Issue Number: https://github.com/apache/incubator-brpc/issues/1992 https://github.com/apache/incubator-brpc/issues/1918

Problem Summary: bRPC as thirdparty deps is not easy.

### What is changed and the side effects?

Changed: I add a example about 1.3 release how to add bRPC to your project and some about this docs.

I have test in my docker debian9 server. bazel version = 4.2.2
and i have test link @apache_brpc:bthread, bvar, butil, brpc all of them are successful.

Side effects:
- Performance effects(性能影响): null

- Breaking backward compatibility(向后兼容性): null

---
### Check List:
- Please make sure your changes are compilable(请确保你的更改可以通过编译).
- When providing us with a new feature, it is best to add related tests(如果你向我们增加一个新的功能, 请添加相关测试).
- Please follow [Contributor Covenant Code of Conduct](../../master/CODE_OF_CONDUCT.md).(请遵循贡献者准则).

### CODE REVIEW

File: example/bazel_1_3_release/WORKSPACE_example [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+)
+
+bind(
+    name = "ssl",
+    actual = "@openssl//:ssl"
+)
+
+
+http_archive(
+    name = "bazel_skylib",
```
Comment (wwbmmm) [inline example/bazel_1_3_release/WORKSPACE_example:88]: duplicated with line 29

File: example/bazel_1_3_release/test/test.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <bthread/execution_queue.h>
+
+int main () {
+  //
```
Comment (wwbmmm) [inline example/bazel_1_3_release/test/test.cpp:21]: you may call some bthread function here

File: docs/cn/bazel_support.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
+## bRPC 作为Bazel第三方依赖
+1. bRPC 依赖于一些开源库, 但这些库并没有提供bazel支持, 所以需要你手动将一部分依赖加入到你的构建项目中.
+2. 将 /example/bazel_1_3_release/build_for_brpc 该文件移动到你的项目根目录下, 将 /example/bazel_1_3_release/build_for_brpc/WORKSPACE_eample 内容添加到你的WORKSPACE中.
```
Comment (wwbmmm) [inline docs/cn/bazel_support.md:3]: you may change the directory name from `bazel_1_3_release` to `build_with_bazel` since there is an existing directory named `build_with_old_bazel`
And you can follow the directory structure of `build_with_old_bazel`

**Comment by fansehep** [comment]
> CI is failed

maybe it is the test failed.

File: docs/cn/bazel_support.md (hunk 3 above)
Comment (fansehep) [reply docs/cn/bazel_support.md]: done.

File: example/bazel_1_3_release/test/test.cpp (hunk 2 above)
Comment (fansehep) [reply example/bazel_1_3_release/test/test.cpp]: > you may call some bthread function here

should we need a test? the BUILD add a empty test here.

File: example/bazel_1_3_release/WORKSPACE_example (hunk 1 above)
Comment (fansehep) [reply example/bazel_1_3_release/WORKSPACE_example]: > Contributor

fixed.

File: example/bazel_1_3_release/WORKSPACE_example [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    ],
+    sha256 = "87407cd28e7a9c95d9f61a098a53cf031109d451a7763e7dd1253abf8b4df422",
+    strip_prefix = "protobuf-3.19.1",
+    urls = ["https://github.com/protocolbuffers/protobuf/archive/refs/tags/v3.19.1.tar.gz"],
+)
+
+
+BAZEL_SKYLIB_VERSION = "1.1.1"  # 2021-09-27T17:33:49Z
+
+BAZEL_SKYLIB_SHA256 = "c6966ec828da198c5d9adbaa94c05e3a1c7f21bd012a0b29ba8ddbccb2c93b0d"
```
[the remark this answers is not in the document]
Comment (fansehep) [reply example/bazel_1_3_release/WORKSPACE_example]: > duplicated with line 25

fixed.

File: example/bazel_1_3_release/WORKSPACE_example [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+Set-Content protobuf.bzl -Value $content -Encoding UTF8
+""",
+    ],
+    sha256 = "87407cd28e7a9c95d9f61a098a53cf031109d451a7763e7dd1253abf8b4df422",
+    strip_prefix = "protobuf-3.19.1",
+    urls = ["https://github.com/protocolbuffers/protobuf/archive/refs/tags/v3.19.1.tar.gz"],
+)
+
+
+BAZEL_SKYLIB_VERSION = "1.1.1"  # 2021-09-27T17:33:49Z
```
[the remark this answers is not in the document]
Comment (fansehep) [reply example/bazel_1_3_release/WORKSPACE_example]: > duplicated with line 23

fixed.

File: docs/en/bazel_support.md [hunk 6]
```diff
@@ -0,0 +1,14 @@
+## bRPC as a Bazel third-party dependency
+1. bRPC relies on a number of open source libraries that do not provide bazel support, so you will need to manually add some of these dependencies to your build project.
+2. Move the file /example/bazel_1_3_release/build_for_brpc to the root of your project, and add the contents of /example/bazel_1_3_release/build_for_brpc/WORKSPACE_eample to your WORKSPACE
```
[the remark this answers is not in the document]
Comment (fansehep) [reply docs/en/bazel_support.md]: > eample -> example

fixed.

File: docs/cn/bazel_support.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (fansehep) [reply docs/cn/bazel_support.md]: > eample -> example

🥲  fixed.

File: docs/cn/bazel_support.md [hunk 7]
```diff
@@ -0,0 +1,14 @@
+## bRPC 作为Bazel第三方依赖
+1. bRPC 依赖于一些开源库, 但这些库并没有提供bazel支持, 所以需要你手动将一部分依赖加入到你的构建项目中.
+2. 将 /example/build_with_bazel/build_for_brpc 该文件移动到你的项目根目录下, 将 /example/build_with_bazel/build_for_brpc/WORKSPACE_example 内容添加到你的WORKSPACE中.
```
Comment (wwbmmm) [inline docs/cn/bazel_support.md:3]: `build_for_brpc/WORKSPACE_example `
this path does not exists now

Comment (fansehep) [reply docs/cn/bazel_support.md]: > 

fixed and add a test for it.

File: docs/cn/bazel_support.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+## bRPC 作为Bazel第三方依赖
+1. bRPC 依赖于一些开源库, 但这些库并没有提供bazel支持, 所以需要你手动将一部分依赖加入到你的构建项目中.
+2. 将 /example/build_with_bazel/build_for_brpc 该文件移动到你的项目根目录下, 将 /example/build_with_bazel/build_for_brpc/WORKSPACE 内容添加到你的WORKSPACE中.
```
Comment (wwbmmm) [inline docs/cn/bazel_support.md:3]: remove `/build_for_brpc`

Comment (fansehep) [reply docs/cn/bazel_support.md]: > remove `/build_for_brpc`

fixed.

File: example/BUILD.bazel [hunk 9]
```diff
@@ -95,8 +95,26 @@ cc_binary(
...
 
+
+bool_flag(
+    name = "rdma",
+    build_setting_default = False,
+    visibility = ["//visibility:public"],
+)
+
+config_setting(
+    name = "have_rdma",
```
Comment (wwbmmm) [inline example/BUILD.bazel:106]: 上面有个 //bazel/config:brpc_with_rdma
看能直接复用吗？

**Comment by wwbmmm** [comment]
LGTM

**Comment by fansehep** [comment]
> Does it support MacOS?

I do not have Mac laptop, you can try it?

**Comment by Huixxi** [comment]
> 
Ok, I will try it later.

**Comment by guodongxiaren** [comment]
> > 
> 
> Ok, I will try it later.
Mac能生效吗？
@Huixxi

File: example/build_with_bazel/brpc_workspace.bzl [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        urls = [
+            "https://downloads.sourceforge.net/project/libpng/zlib/1.2.11/zlib-1.2.11.tar.gz",
+            "https://zlib.net/fossils/zlib-1.2.11.tar.gz",
+        ],
+    )
+
+    native.new_local_repository(
+        name = "openssl",
+        path = "/usr",
+        build_file = "//:openssl.BUILD",
```
Comment (372046933) [inline example/build_with_bazel/brpc_workspace.bzl:80]: `//:openssl.BUILD` actually links to openssl on the system. While openssl in BRPC  WORKSPACE is compiled by foreign_rules.
It's better to keep the same with that in BRPC WORKSPACE
So brpc_workspace.bzl is different from the original `WORKSPACE` file. Could `brpc_workspace` be used by `WORKSPACE` in BRPC as well to maintain a consistent `WORKSPACE`?

**Review by 372046933 — COMMENTED** [review]
Could `brpc_workspace` be applied to BRPC WORKSPACE?

**Comment by fansehep** [comment]
> > 
> 
> Ok, I will try it later.

@Huixxi Have you test successful in ```MacOS```?
Here is a simple demo in this pull request.
```bash
cd example/build_with_bazel
bazel build "..."
./bazel-bin/test
```
Try this?

File: example/build_with_bazel/brpc_workspace.bzl [hunk 11]
```diff
@@ -74,10 +74,15 @@ def brpc_workspace():
         ],
     )
 
-    native.new_local_repository(
-        name = "openssl",
-        path = "/usr",
+    http_archive(
+        name = "openssl",  # 2021-12-14T15:45:01Z
```
Comment (372046933) [inline example/build_with_bazel/brpc_workspace.bzl:78]: `example/build_with_bazel/openssl.BUILD` does not build but links to system installed openssl.
Here is the real build:
https://github.com/apache/incubator-brpc/blob/fae30e6a9a84ef8c2e0bfe1ce23afa5611a1ffab/bazel/third_party/openssl/openssl.BUILD#L132

**Comment by zyearn** [comment]
Hi, What's the status on it? Could you update the latest code?

**Comment by fansehep** [comment]
> Hi, What's the status on it? Could you update the latest code?

In fact, this is ok. I have tried some other methods, but none of them seem very elegant. So I thinks it can be merged. 😢

**Comment by 89dexua** [comment]
Any chance to support protobuf 3.21.9
Tensorflow 2.15.0 not Compatible with protobuf 3.19.1

Simply change the protobuf version got this error message 

ERROR: /home/hitjinzhou/brpc/example/BUILD.bazel:40:14: every rule of type proto_library implicitly depends upon the target '@com_google_protobuf//:protoc', but this target could not be found because of: error loading package '@com_google_protobuf//': cannot load '@com_google_protobuf//:compiler_config_setting.bzl': no such file

**Comment by fansehep** [comment]
> Any chance to support protobuf 3.21.9 Tensorflow 2.15.0 not Compatible with protobuf 3.19.1
> 
> Simply change the protobuf version got this error message
> 
> ERROR: brpc/example/BUILD.bazel:40:14: every rule of type proto_library implicitly depends upon the target '@com_google_protobuf//:protoc', but this target could not be found because of: error loading package '@com_google_protobuf//': cannot load '@com_google_protobuf//:compiler_config_setting.bzl': no such file

`Problems with rhombic dependence`, Can try change the build stricp brpc protobuf name to fix it.
